### PR TITLE
Improve portability of libffi-sys

### PR DESCRIPTION
**Pull request by AndrewGaspar**
This change attempts to make libffi-sys more portable to platforms other than `x86_64-unknown-linux-gnu` through a careful reading of libffi, Rust, and LLVM.

The constants which are redefined in this PR include:
- ffi_abi_*
- FFI_TRAMPOLINE_SIZE
- FFI_NATIVE_RAW_API
- FFI_GO_CLOSURES

This almost certainly isn't exhaustive - there are probably other divergences.

I've tested on:
- x86_64-pc-windows-msvc
- i686-pc-windows-msvc
- x86_64-unknown-linux-gnu (features=system)

I made the choice to avoid masking out large sections of code with `#[cfg]` to hopefully catch more issues at compile time. Instead, I try to only use `#[cfg]` to actually publish the host names.

Unfortunately, I don't have resources on-hand to test aarch64 or powerpc64. I wanted to check cross-compilation, but I couldn't get the code to build without features=system in my Ubuntu install. I think something about autoconf not being able to find `ld`? I'm not sure, the error message was difficult to understand.

**EDIT:** libffi-rs is now tested on:
* `x86_64-unknown-linux-gnu`
* `i686-unknown-linux-gnu`
* `aarch64-unknown-linux-gnu`
- `armv7-unknown-linux-gnueabihf`
* `powerpc64-unknown-linux-gnu`
* `powerpc64le-unknown-linux-gnu`

NOTE: `libffi-sys-rs` does not currently build for `powerpc-unknown-linux-gnu` with libffi 3.3, but it does build if you update libffi to the latest master branch. Unfortunately, the tests do not currently pass. Somebody else may have to figure out what the bug is.

### CODE REVIEW

File: libffi-sys-rs/src/arch.rs [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+        pub const ffi_abi_FFI_DEFAULT_ABI: ffi_abi =
+            ffi_abi_FFI_LINUX | STRUCT_ALIGN_FLAG | LONG_DOUBLE_128_FLAG;
+
+        pub const FFI_NATIVE_RAW_API: u32 = 0;
+        pub const FFI_GO_CLOSURES: u32 = 1;
+
+        // If no elfv2...
+        #[cfg(all(target_arch = "powerpc64", not(target_env = "musl")))]
+        pub const FFI_TRAMPOLINE_SIZE: usize = 24;
```
Comment (AndrewGaspar) [inline libffi-sys-rs/src/arch.rs:286]: Hm... I don't know why I didn't just stick this into `mod elfv1`

Comment (AndrewGaspar) [reply libffi-sys-rs/src/arch.rs:286]: I need to fix something with this.

**Comment by timfish** [comment]
I'm tempted to offer a couple of Raspberry Pi's which we can setup with GitHub actions as [self hosted runners](https://docs.github.com/en/free-pro-team@latest/actions/hosting-your-own-runners/using-self-hosted-runners-in-a-workflow). 

That would allow us to have all the tests running on `aarch64` and `armv7` too. My main concern would be that it opens up my network to abuse from malicious PRs 🤔. I _could_ setup some sort of network segmentation but the equipment I currently have in my home network does not support this.

**Comment by timfish** [comment]
Looks like we could probably get `aarch64` and `armv7` working via `cross`, although [it looks like](https://github.com/tov/libffi-rs/pull/18) the base docker images don't have the required dependencies so we'd need some custom images. I'll look into this when I get some more time.

**Comment by YorickPeterse** [comment]
@timfish I think setting up Raspberry Pi's for testing may be a bit too risky. And I think they wouldn't be used for forks anyway (unless GitHub does support custom runners for forks these days), so they wouldn't be terribly useful.

Regarding dependencies, if we can install libffi in the mentioned Docker image we can adjust the test matrix so only tests using the system FFI are run using this image. This may make it easier to set things up.

A possible hack would be to use QEMU or similar to run an ARM installation inside whatever GitHub actions use. I however suspect this is going to be quite slow, and possibly painful to maintain.

**Comment by AndrewGaspar** [comment]
Yeah, I was thinking maybe we could use QEMU. We should be able to cross-compile on the native architecture and then copy the test executables into the QEMU image, which I would think would be fairly fast.

**Comment by AndrewGaspar** [comment]
Yeah, qemu is working for me. In my experience it's more than fast enough, as long as you're cross-compiling from x86. I figured out why my default build of `x86_64-pc-windows-msvc` wasn't working, too.

I have some changes in-coming to:
1) fix powerpc64le-unknown-linux-gnu
2) Add tests that use qemu to run aarch64 and powerpc64 tests

**Comment by AndrewGaspar** [comment]
I was able to get the tests to pass on:
1) powerpc64le-unknown-linux-gnu
2) powerpc64-unknown-linux-gnu
3) aarch64-unknown-linux-gnu
4) i686-unknown-linux-gnu

all through QEMU, cross-compiled from the host. I'm still trying to get 32-bit powerpc passing, but it's proving difficult, so I might abandon it, tbh. Not sure anybody cares about powerpc.

**Comment by timfish** [comment]
Any chance you can add `armv7-unknown-linux-gnueabihf`?

**Comment by AndrewGaspar** [comment]
> And how easy would it be to get those tests running on Github actions with the existing tests?

It should be easy - I'm planning to do that before taking away WIP.

**Comment by AndrewGaspar** [comment]
> Any chance you can check armv7-unknown-linux-gnueabihf?

Passes for me!

File: libffi-sys-rs/build/not_msvc.rs [hunk 2]
```diff
@@ -50,8 +52,28 @@ pub fn configure_libffi(prefix: PathBuf, build_dir: &Path) {
...
+            command.arg("--enable-debug");
+        }
+    }
+
+    let target = std::env::var("TARGET").unwrap();
+    if target != std::env::var("HOST").unwrap() {
+        command.arg(format!("--host={}", target.to_string()));
+    }
+
+    command.current_dir(&build_dir);
```
Comment (YorickPeterse) [inline libffi-sys-rs/build/not_msvc.rs:76]: Are these changes really necessary? I'd rather not introduce additional dependencies _just_ so we can enable debug assertions.

File: libffi-sys-rs/src/lib.rs [hunk 3]
```diff
@@ -162,17 +177,31 @@ pub struct ffi_closure {
...
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ffi_closure")
+            .field("tramp", &&self.tramp[..])
+            .field("cif", &self.cif)
+            .field("fun", &self.fun)
+            .field("user_data", &self.user_data)
+            .finish()
+    }
+}
+
```
Comment (YorickPeterse) [inline libffi-sys-rs/src/lib.rs:191]: Is this really necessary in this PR? If the intentions are to improve portability, I'd rather have the PR focus on _just_ that; anything not directly related to that should go in a separate PR, making it easier to review changes :smile:

File: libffi-sys-rs/build/not_msvc.rs (hunk 2 above)
Comment (AndrewGaspar) [reply libffi-sys-rs/build/not_msvc.rs:76]: No, the debug assertions aren't necessary. They DID catch a bug and prevented undefined behavior when targeting `powerpc-unknown-linux-gnu`, though I haven't been able to fix the bug yet.

I can remove if you'd like.

File: libffi-sys-rs/src/lib.rs (hunk 3 above)
Comment (AndrewGaspar) [reply libffi-sys-rs/src/lib.rs:191]: It is related to portability - it lets the code build on `powerpc-unknown-linux-gnu`, though that target has other problems.

File: libffi-sys-rs/build/not_msvc.rs (hunk 2 above)
Comment (YorickPeterse) [reply libffi-sys-rs/build/not_msvc.rs:76]: Since it's not directly related to this PR I'd prefer to not include it. I'm fine with including it in a separate PR, but it would be nice if we could do so without adding any additional (build time) dependencies.

File: libffi-sys-rs/src/lib.rs (hunk 3 above)
Comment (YorickPeterse) [reply libffi-sys-rs/src/lib.rs:191]: In that case, how does implementing Debug/etc make this code more portable? Is it because we're already deriving `Debug` in places and that doesn't work on powerpc?

Comment (AndrewGaspar) [reply libffi-sys-rs/src/lib.rs:191]: You can't `derive(Debug)` on structs with arrays that have more than 32 elements. On PowerPC `FFI_TRAMPOLINE_SIZE` is 40.

Comment (YorickPeterse) [reply libffi-sys-rs/src/lib.rs:191]: Aaah gotcha, then we can keep the Debug implementation code. Thanks for clarifying! :smile:

**Comment by AndrewGaspar** [comment]
Should be good now.

**Comment by timfish** [comment]
This is great. 48 platform/arch combinations compiled and tested in 4m 16s! 

👍

**Comment by timfish** [comment]
@YorickPeterse when you get some time, could we have this merged and a crate release?

Many thanks!

**Comment by YorickPeterse** [comment]
@AndrewGaspar Thanks!

**Comment by AndrewGaspar** [comment]
Will 1.1.0 be published to crates.io soon?

**Comment by YorickPeterse** [comment]
@AndrewGaspar It was published on November 11: https://crates.io/crates/libffi-sys/1.1.0

**Comment by AndrewGaspar** [comment]
D'oh, for some reason I was looking for a version bump for libffi. Thanks!

**Comment by timfish** [comment]
Yep, I did exactly the same!
